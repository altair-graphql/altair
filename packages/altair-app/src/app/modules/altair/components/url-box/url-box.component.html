<div class="url-box">
  <div class="url-box__method-wrapper">
    <button
      type="button"
      nz-dropdown
      nzTrigger="click"
      [nzDropdownMenu]="httpVerbMenu"
      class="url-box__button url-box__button--method"
      [value]="httpVerb"
      track-id="http_verb"
    >
      {{ httpVerb }}
      <!-- &#9660; -->
    </button>

    <nz-dropdown-menu #httpVerbMenu="nzDropdownMenu">
      <ul nz-menu>
        @for (method of methods; track method) {
          <li nz-menu-item [value]="method" (click)="setVerb(method)">
            {{ method }}
          </li>
        }
      </ul>
    </nz-dropdown-menu>
  </div>
  <div class="url-box__input-container">
    @if (streamState) {
      <div class="url-box__input-prepend">
        <div class="url-box__connection-indicator">
          @if (streamState === 'uncertain') {
            <app-icon
              name="alert-circle"
              class="connection-indicator--uncertain"
              nz-tooltip
              [nzTooltipTitle]="'STREAM_CONNECTION_TRYING' | translate"
              [nzTooltipOverlayStyle]="{ 'white-space': 'nowrap' }"
            ></app-icon>
          }
          @if (streamState === 'connected') {
            <app-icon
              name="check-circle"
              class="connection-indicator--connected"
              nz-tooltip
              [nzTooltipTitle]="'STREAM_CONNECTION_CONNECTED' | translate"
              [nzTooltipOverlayStyle]="{ 'white-space': 'nowrap' }"
            ></app-icon>
          }
        </div>
      </div>
    }
    <app-x-input
      class="url-box__input"
      track-id="set_url"
      placeholder="Enter URL"
      [(ngModel)]="apiUrl"
      (blurChange)="setApiUrl()"
      (submitChange)="setApiUrl()"
    ></app-x-input>
  </div>
  <div class="url-box__input-actions">
    @if (currentCollection) {
      <button
        type="button"
        class="url-box__button"
        [ngClass]="{ 'url-box__button--disabled': !hasUnsavedChanges }"
        (click)="updateQueryInCollectionChange.emit()"
        track-id="update_in_collection"
        nz-tooltip
        [nzTooltipTitle]="'COLLECTION_UPDATE_WINDOW_IN_COLLECTION' | translate"
        nzTooltipPlacement="bottom"
      >
        <app-icon name="save"></app-icon>
      </button>
    }
    <button
      type="button"
      class="url-box__button"
      (click)="addToCollectionChange.emit()"
      track-id="save_to_collection"
      nz-tooltip
      [nzTooltipTitle]="'SAVE_TO_COLLECTION_BUTTON' | translate"
      nzTooltipPlacement="bottom"
    >
      <app-icon name="plus-circle"></app-icon>
    </button>
    <button
      type="button"
      class="url-box__button"
      (click)="exportWindowChange.emit()"
      track-id="export_window"
      nz-tooltip
      [nzTooltipTitle]="'EXPORT_WINDOW_TEXT' | translate"
      nzTooltipPlacement="bottom"
    >
      <app-icon name="file-down"></app-icon>
    </button>
    <button
      type="button"
      class="url-box__button"
      (click)="reloadDocsChange.emit()"
      track-id="reload_docs"
      nz-tooltip
      [nzTooltipTitle]="'RELOAD_DOCS_BUTTON' | translate"
      nzTooltipPlacement="bottom"
    >
      <app-icon name="refresh-ccw"></app-icon>
    </button>
    <button
      type="button"
      class="url-box__button"
      (click)="toggleDocsChange.emit()"
      [ngClass]="{ 'url-box__button--active': showDocs }"
      track-id="show_docs"
    >
      {{ 'DOCS_TEXT' | translate }}
    </button>
  </div>
  <div class="send-request__wrapper">
    @if (queryOperations.length < 2) {
      <button
        class="app-button active-primary send-request__button"
        (click)="!isSubscribed && this.sendRequest.emit()"
        [ngClass]="{ 'url-box__button--disabled': isSubscribed || isLoading }"
        track-id="set_request"
      >
        @if (isLoading) {
          <span>{{ 'LOADING_INDICATOR_TEXT' | translate }}</span>
        }
        @if (!isLoading) {
          <span>{{ 'SEND_REQUEST_BUTTON' | translate }}</span>
        }
      </button>
    }
    @if (queryOperations.length > 1) {
      <button
        class="app-button active-primary send-request__button"
        [ngClass]="{ 'url-box__button--disabled': isSubscribed }"
        track-id="send_request"
        nz-dropdown
        nzTrigger="click"
        nzPlacement="bottomRight"
        [nzDropdownMenu]="queryOperationsMenu"
      >
        @if (isLoading) {
          <span>{{ 'LOADING_INDICATOR_TEXT' | translate }}</span>
        } @else {
          @if (selectedOperation) {
            <span>{{ 'SEND_REQUEST_BUTTON_SHORT' | translate }}</span>
          } @else {
            <span>{{ 'SEND_REQUEST_BUTTON' | translate }}</span>
          }
        }
        @if (selectedOperation === BATCHED_REQUESTS_OPERATION) {
          <span>(Batched)</span>
        } @else if (selectedOperation) {
          <span [title]="selectedOperation"
            >({{ selectedOperation | truncate: 7 }})</span
          >
        }
        <app-icon name="chevron-down"></app-icon>
      </button>
    }
    <nz-dropdown-menu #queryOperationsMenu="nzDropdownMenu">
      <ul nz-menu>
        @for (
          operation of queryOperations;
          track queryOperationTrackBy($index, operation)
        ) {
          <li
            nz-menu-item
            (click)="selectedOperationChange.emit(operation.name?.value)"
          >
            {{ operation.name?.value }}
          </li>
        }
        <li
          nz-menu-item
          (click)="selectedOperationChange.emit(BATCHED_REQUESTS_OPERATION)"
        >
          {{ 'SEND_BATCHED_REQUEST' | translate }}
        </li>
      </ul>
    </nz-dropdown-menu>
  </div>
</div>
@if (isLoading) {
  <div class="url-box__loader">
    <div class="url-box__loader-inner"></div>
  </div>
}
