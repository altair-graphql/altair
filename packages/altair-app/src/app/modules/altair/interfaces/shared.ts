export interface IDictionary<V = any> {
  [key: string]: V;
}

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type TODO = any;

export type UnknownError = TODO;

export interface TrackByIdItem {
  id: string;
}
