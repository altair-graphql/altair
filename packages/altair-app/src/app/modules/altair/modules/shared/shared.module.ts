import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';

import { DragDropModule } from '@angular/cdk/drag-drop';
import { TranslateModule } from '@ngx-translate/core';
import { NgPipesModule } from 'ngx-pipes';
import { NguiAutoCompleteModule } from '@ngui/auto-complete';
import { MarkdownModule } from 'ngx-markdown';
import { HotToastModule } from '@ngneat/hot-toast';
import { ResizableModule } from 'angular-resizable-element';

import { NzButtonModule } from 'ng-zorro-antd/button';
import { NzFormModule } from 'ng-zorro-antd/form';
import { NzSelectModule } from 'ng-zorro-antd/select';
import { NzModalModule } from 'ng-zorro-antd/modal';
import { NzDropDownModule } from 'ng-zorro-antd/dropdown';
import { NzSwitchModule } from 'ng-zorro-antd/switch';
import { NzListModule } from 'ng-zorro-antd/list';
import { NzRadioModule } from 'ng-zorro-antd/radio';
import { NzAlertModule } from 'ng-zorro-antd/alert';
import { NzLayoutModule } from 'ng-zorro-antd/layout';
import { NzMenuModule } from 'ng-zorro-antd/menu';
import { NzCollapseModule } from 'ng-zorro-antd/collapse';
import { NzInputModule } from 'ng-zorro-antd/input';
import { NzInputNumberModule } from 'ng-zorro-antd/input-number';
import { NzAutocompleteModule } from 'ng-zorro-antd/auto-complete';
import { NzCheckboxModule } from 'ng-zorro-antd/checkbox';
import { NzToolTipModule } from 'ng-zorro-antd/tooltip';
import { NzTabsModule } from 'ng-zorro-antd/tabs';
import { NzTableModule } from 'ng-zorro-antd/table';
import { NzTreeSelectModule } from 'ng-zorro-antd/tree-select';
import { NzGridModule } from 'ng-zorro-antd/grid';
import { NzPopoverModule } from 'ng-zorro-antd/popover';

import { IconsModule } from '../icons/icons.module';
import { NzAvatarModule } from 'ng-zorro-antd/avatar';

const AntdComponentModules = [
  NzButtonModule,
  NzInputModule,
  NzInputNumberModule,
  NzAutocompleteModule,
  NzCheckboxModule,
  NzFormModule,
  NzGridModule,
  NzTabsModule,
  NzSelectModule,
  NzModalModule,
  NzDropDownModule,
  NzSwitchModule,
  NzListModule,
  NzRadioModule,
  NzAlertModule,
  NzLayoutModule,
  NzMenuModule,
  NzCollapseModule,
  NzToolTipModule,
  NzTableModule,
  NzTreeSelectModule,
  NzPopoverModule,
  NzAvatarModule,
  // NgZorroAntdModule,
];

@NgModule({
  imports: [
    CommonModule,
    DragDropModule,
    TranslateModule.forRoot(),
    NgPipesModule,
    NguiAutoCompleteModule, // use antd
    MarkdownModule.forRoot(),
    ResizableModule,
    IconsModule,
    // ToastrModule.forRoot({
    //   newestOnTop: false,
    //   closeButton: true,
    //   positionClass: 'toast-top-center',
    //   enableHtml: true,
    //   countDuplicates: true,
    //   preventDuplicates: true,
    //   resetTimeoutOnDuplicate: true,
    // }),
    // HotToastModule.forRoot({
    //   position: 'top-center',
    //   reverseOrder: true,
    //   dismissible: true,
    //   closeStyle: {
    //     backgroundImage: `none`,
    //   },
    // }),
    ...AntdComponentModules,
  ],

  // DON'T FORGET TO EXPORT MODULE
  exports: [
    DragDropModule,
    TranslateModule,
    NgPipesModule,
    NguiAutoCompleteModule,
    MarkdownModule,
    ResizableModule,
    IconsModule,
    HotToastModule,
    ...AntdComponentModules,
  ],
})
export class SharedModule {
  static forRoot(): ModuleWithProviders<SharedModule> {
    return {
      ngModule: SharedModule,
      providers: [],
    };
  }
}
