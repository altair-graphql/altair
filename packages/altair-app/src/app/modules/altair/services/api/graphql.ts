import { gql } from 'apollo-angular';
import { Injectable } from '@angular/core';
import * as Apollo from 'apollo-angular';
import * as ApolloCore from '@apollo/client/core';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type CreateRequestCollectionInput = {
  name: Scalars['String'];
  workspaceId?: InputMaybe<Scalars['String']>;
};

export type CreateRequestInput = {
  collectionId?: InputMaybe<Scalars['String']>;
  headers?: InputMaybe<Array<InputMaybe<HeaderInput>>>;
  name?: InputMaybe<Scalars['String']>;
  ownerId?: InputMaybe<Scalars['String']>;
  query?: InputMaybe<Scalars['String']>;
  url?: InputMaybe<Scalars['String']>;
  variables?: InputMaybe<Scalars['String']>;
};

export type CreateUserInput = {
  email: Scalars['String'];
  firstName: Scalars['String'];
  lastName?: InputMaybe<Scalars['String']>;
};

export type CreateWorkspaceInput = {
  name: Scalars['String'];
};

export type Header = {
  __typename?: 'Header';
  key: Scalars['String'];
  value: Scalars['String'];
};

export type HeaderInput = {
  key: Scalars['String'];
  value: Scalars['String'];
};

export type Mutation = {
  __typename?: 'Mutation';
  createRequest: Request;
  createRequestCollection: RequestCollection;
  createWorkspace: Workspace;
  getOrCreateUser?: Maybe<User>;
  removeRequest?: Maybe<Request>;
  removeRequestCollection?: Maybe<RequestCollection>;
  removeWorkspace?: Maybe<Workspace>;
  updateRequest: Request;
  updateRequestCollection: RequestCollection;
  updateWorkspace: Workspace;
};


export type MutationCreateRequestArgs = {
  createRequestInput: CreateRequestInput;
};


export type MutationCreateRequestCollectionArgs = {
  createRequestCollectionInput: CreateRequestCollectionInput;
};


export type MutationCreateWorkspaceArgs = {
  createWorkspaceInput: CreateWorkspaceInput;
};


export type MutationRemoveRequestArgs = {
  id: Scalars['Int'];
};


export type MutationRemoveRequestCollectionArgs = {
  id: Scalars['Int'];
};


export type MutationRemoveWorkspaceArgs = {
  id: Scalars['String'];
};


export type MutationUpdateRequestArgs = {
  updateRequestInput: UpdateRequestInput;
};


export type MutationUpdateRequestCollectionArgs = {
  updateRequestCollectionInput: UpdateRequestCollectionInput;
};


export type MutationUpdateWorkspaceArgs = {
  updateWorkspaceInput: UpdateWorkspaceInput;
};

export type Query = {
  __typename?: 'Query';
  getUserById?: Maybe<User>;
  profile?: Maybe<User>;
  request?: Maybe<Request>;
  requestCollection?: Maybe<RequestCollection>;
  requestCollections: Array<Maybe<RequestCollection>>;
  requests: Array<Maybe<Request>>;
  workspace?: Maybe<Workspace>;
  workspaces: Array<Maybe<Workspace>>;
};


export type QueryGetUserByIdArgs = {
  id: Scalars['String'];
};


export type QueryRequestArgs = {
  id: Scalars['Int'];
};


export type QueryRequestCollectionArgs = {
  id: Scalars['Int'];
};


export type QueryWorkspaceArgs = {
  id: Scalars['Int'];
};

export type Request = {
  __typename?: 'Request';
  collectionId?: Maybe<Scalars['String']>;
  headers?: Maybe<Array<Maybe<Header>>>;
  name?: Maybe<Scalars['String']>;
  ownerId?: Maybe<Scalars['String']>;
  query?: Maybe<Scalars['String']>;
  url?: Maybe<Scalars['String']>;
  variables?: Maybe<Scalars['String']>;
};

export type RequestCollection = {
  __typename?: 'RequestCollection';
  name?: Maybe<Scalars['String']>;
  workspaceId?: Maybe<Scalars['String']>;
};

export type UpdateRequestCollectionInput = {
  id: Scalars['Int'];
};

export type UpdateRequestInput = {
  id: Scalars['Int'];
};

export type UpdateWorkspaceInput = {
  id: Scalars['String'];
  name?: InputMaybe<Scalars['String']>;
};

export type User = {
  __typename?: 'User';
  email: Scalars['String'];
  firstName: Scalars['String'];
  id: Scalars['String'];
  lastName?: Maybe<Scalars['String']>;
};

export type Workspace = {
  __typename?: 'Workspace';
  id?: Maybe<Scalars['String']>;
  name?: Maybe<Scalars['String']>;
};

export type ProfileQueryVariables = Exact<{ [key: string]: never; }>;


export type ProfileQuery = { __typename?: 'Query', profile?: { __typename?: 'User', email: string, firstName: string, lastName?: string | null | undefined } | null | undefined };

export type GetOrCreateUserMutationVariables = Exact<{ [key: string]: never; }>;


export type GetOrCreateUserMutation = { __typename?: 'Mutation', getOrCreateUser?: { __typename?: 'User', email: string, firstName: string, lastName?: string | null | undefined } | null | undefined };

export const ProfileDocument = gql`
    query Profile {
  profile {
    email
    firstName
    lastName
  }
}
    `;

  @Injectable({
    providedIn: 'root'
  })
  export class ProfileGQL extends Apollo.Query<ProfileQuery, ProfileQueryVariables> {
    document = ProfileDocument;
    
    constructor(apollo: Apollo.Apollo) {
      super(apollo);
    }
  }
export const GetOrCreateUserDocument = gql`
    mutation getOrCreateUser {
  getOrCreateUser {
    email
    firstName
    lastName
  }
}
    `;

  @Injectable({
    providedIn: 'root'
  })
  export class GetOrCreateUserGQL extends Apollo.Mutation<GetOrCreateUserMutation, GetOrCreateUserMutationVariables> {
    document = GetOrCreateUserDocument;
    
    constructor(apollo: Apollo.Apollo) {
      super(apollo);
    }
  }

  type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

  interface WatchQueryOptionsAlone<V>
    extends Omit<ApolloCore.WatchQueryOptions<V>, 'query' | 'variables'> {}
    
  interface QueryOptionsAlone<V>
    extends Omit<ApolloCore.QueryOptions<V>, 'query' | 'variables'> {}
    
  interface MutationOptionsAlone<T, V>
    extends Omit<ApolloCore.MutationOptions<T, V>, 'mutation' | 'variables'> {}
    
  interface SubscriptionOptionsAlone<V>
    extends Omit<ApolloCore.SubscriptionOptions<V>, 'query' | 'variables'> {}

  @Injectable({ providedIn: 'root' })
  export class APIClient {
    constructor(
      private profileGql: ProfileGQL,
      private getOrCreateUserGql: GetOrCreateUserGQL
    ) {}
      
    profile(variables?: ProfileQueryVariables, options?: QueryOptionsAlone<ProfileQueryVariables>) {
      return this.profileGql.fetch(variables, options)
    }
    
    profileWatch(variables?: ProfileQueryVariables, options?: WatchQueryOptionsAlone<ProfileQueryVariables>) {
      return this.profileGql.watch(variables, options)
    }
    
    getOrCreateUser(variables?: GetOrCreateUserMutationVariables, options?: MutationOptionsAlone<GetOrCreateUserMutation, GetOrCreateUserMutationVariables>) {
      return this.getOrCreateUserGql.mutate(variables, options)
    }
  }