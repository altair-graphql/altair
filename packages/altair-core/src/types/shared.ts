export type IDictionary<V = any> = Record<string, V>;

export type TODO = any;

// export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export interface BaseDocument {
  id: string;
  created_at?: number;
  updated_at?: number;
}

export interface BaseOwnableDocument extends BaseDocument {
  ownerUid: string;
}

type TypedOmit<T, K extends keyof T> = Omit<T, K>;
export type CreateDTO<
  T extends BaseDocument | BaseOwnableDocument
> = T extends BaseOwnableDocument
  ? TypedOmit<T, keyof BaseOwnableDocument>
  : TypedOmit<T, keyof BaseDocument>;
export type UpdateDTO<T extends BaseDocument | BaseOwnableDocument> = CreateDTO<
  T
> &
  Pick<T, 'id'>;
