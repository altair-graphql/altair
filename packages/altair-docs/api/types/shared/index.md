# types/shared

## Index

| Member | Description |
| :------ | :------ |
| [BaseDocument](interfaces/BaseDocument.md) | - |
| [BaseOwnableDocument](interfaces/BaseOwnableDocument.md) | - |
| [CreateDTO](type-aliases/CreateDTO.md) | - |
| [IDictionary](type-aliases/IDictionary.md) | - |
| [NoExtraProperties](type-aliases/NoExtraProperties.md) | - |
| [TODO](type-aliases/TODO.md) | - |
| [UpdateDTO](type-aliases/UpdateDTO.md) | - |
