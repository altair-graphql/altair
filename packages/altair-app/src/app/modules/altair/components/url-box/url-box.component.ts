import { Component, EventEmitter, Input, Output } from '@angular/core';
import { IQueryCollection } from 'altair-graphql-core/build/types/state/collection.interfaces';
import { HTTP_VERBS } from 'altair-graphql-core/build/types/state/query.interfaces';
import { OperationDefinitionNode } from 'graphql';
import { VARIABLE_REGEX } from '../../services/environment/environment.service';
import { BATCHED_REQUESTS_OPERATION } from '../../services/gql/gql.service';

@Component({
  selector: 'app-url-box',
  templateUrl: './url-box.component.html',
})
export class UrlBoxComponent {
  @Input() apiUrl = '';
  @Input() httpVerb = 'POST';
  @Input() isSubscribed = false;
  @Input() isLoading = false;
  @Input() showDocs = false;
  @Input() selectedOperation = '';
  @Input() queryOperations: OperationDefinitionNode[] = [];
  @Input() streamState = '';
  @Input() currentCollection?: IQueryCollection;
  @Input() hasUnsavedChanges = false;

  @Output() toggleDocsChange = new EventEmitter();
  @Output() reloadDocsChange = new EventEmitter();
  @Output() addToCollectionChange = new EventEmitter();
  @Output() sendRequest = new EventEmitter();
  @Output() urlChange = new EventEmitter();
  @Output() httpVerbChange = new EventEmitter();
  @Output() selectedOperationChange = new EventEmitter();
  @Output() exportWindowChange = new EventEmitter();
  @Output() updateQueryInCollectionChange = new EventEmitter();

  methods = HTTP_VERBS;

  BATCHED_REQUESTS_OPERATION = BATCHED_REQUESTS_OPERATION;

  setApiUrl() {
    const sanitizedUrl = this.sanitizeUrl(this.apiUrl);
    this.urlChange.emit(sanitizedUrl);
  }

  setVerb(verb: string) {
    this.httpVerbChange.emit(verb);
  }

  sanitizeUrl(url: string) {
    // trim the url and remove any spaces
    // add http protocol if missing
    url = url.trim();
    const hasProtocol = /^[a-zA-Z]+:\/\//.test(url);
    const hasVariable = /{{.*}}/.test(url);
    if (!(hasProtocol || hasVariable)) {
      return 'http://' + url;
    }

    return url;
  }

  queryOperationTrackBy(index: number, operation: OperationDefinitionNode) {
    return operation.name?.value;
  }
}
