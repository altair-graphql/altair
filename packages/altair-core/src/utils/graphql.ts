import { DocumentNode, Kind, OperationDefinitionNode, parse } from 'graphql';

const getEmptyDocumentNode = (): DocumentNode => {
  return {
    definitions: [],
    kind: Kind.DOCUMENT,
  };
};
const parseQueryOrEmptyDocument = (query: string) => {
  if (!query) {
    return getEmptyDocumentNode();
  }

  try {
    return parse(query);
  } catch (err) {
    console.error('Error parsing query, returning empty instead', err);

    return getEmptyDocumentNode();
  }
};

export const getOperations = (query: string) => {
  const parsedQuery = parseQueryOrEmptyDocument(query);

  if (parsedQuery.definitions) {
    return parsedQuery.definitions.filter(
      (def): def is OperationDefinitionNode =>
        !!(def.kind === 'OperationDefinition' && def.name?.value)
    );
  }

  return [];
};
