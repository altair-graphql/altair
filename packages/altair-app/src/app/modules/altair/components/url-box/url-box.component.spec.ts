import { expect, describe, it } from '@jest/globals';

import { UrlBoxComponent } from './url-box.component';

import { SharedModule } from '../../modules/shared/shared.module';
import { NgxTestWrapper } from '../../../../../testing/wrapper';
import { mount } from '../../../../../testing';
import { MockModule } from 'ng-mocks';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { OperationDefinitionNode } from 'graphql';

describe('UrlBoxComponent', () => {
  let wrapper: NgxTestWrapper<UrlBoxComponent>;

  beforeEach(async () => {
    wrapper = await mount({
      component: UrlBoxComponent,
      imports: [MockModule(SharedModule)],
      schemas: [NO_ERRORS_SCHEMA],
    });
  });

  it('should create', () => {
    expect(wrapper.componentInstance).toBeTruthy();
  });

  it('should render correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render correctly with queryOperations > 1', async () => {
    wrapper.setProps({
      queryOperations: [
        {
          name: { value: 'operation 1', kind: 'Name' },
        } as OperationDefinitionNode,
        {
          name: { value: 'operation 2', kind: 'Name' },
        } as OperationDefinitionNode,
      ],
    });
    await wrapper.nextTick();
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('sanitizeUrl', () => {
    it('should add http protocol if missing', () => {
      expect(wrapper.componentInstance.sanitizeUrl('localhost:3000')).toBe(
        'http://localhost:3000'
      );
    });

    it('should not add http protocol if already present', () => {
      expect(
        wrapper.componentInstance.sanitizeUrl('http://localhost:3000')
      ).toBe('http://localhost:3000');
    });

    it('should not add http protocol if url contains https', () => {
      expect(
        wrapper.componentInstance.sanitizeUrl('https://localhost:3000')
      ).toBe('https://localhost:3000');
    });

    it('should not add protocol if url contains variable', () => {
      expect(wrapper.componentInstance.sanitizeUrl('{{url}}/graphql')).toBe(
        '{{url}}/graphql'
      );
    });
  });
});
