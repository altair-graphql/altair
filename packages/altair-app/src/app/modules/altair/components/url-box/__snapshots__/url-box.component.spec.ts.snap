// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UrlBoxComponent should render correctly 1`] = `
<app-url-box>
  <div
    class="url-box"
  >
    <div
      class="url-box__method-wrapper"
    >
      <button
        class="url-box__button url-box__button--method"
        nz-dropdown=""
        nztrigger="click"
        track-id="http_verb"
        type="button"
        value="POST"
      >
         POST 
      </button>
      
      <nz-dropdown-menu>
        <ul
          nz-menu=""
        >
          <li
            nz-menu-item=""
            value="0"
          >
             POST 
          </li>
          
          <li
            nz-menu-item=""
            value="0"
          >
             GET 
          </li>
          
          <li
            nz-menu-item=""
            value="0"
          >
             PUT 
          </li>
          
          <li
            nz-menu-item=""
            value="0"
          >
             DELETE 
          </li>
          
          
        </ul>
        
      </nz-dropdown-menu>
    </div>
    <div
      class="url-box__input-container"
    >
      
      <app-x-input
        class="url-box__input"
        placeholder="Enter URL"
        track-id="set_url"
      />
      
    </div>
    <div
      class="url-box__input-actions"
    >
      
      <button
        class="url-box__button"
        nz-tooltip=""
        nztooltipplacement="bottom"
        track-id="save_to_collection"
        type="button"
      >
        <app-icon
          name="plus-circle"
        />
      </button>
      
      <button
        class="url-box__button"
        nz-tooltip=""
        nztooltipplacement="bottom"
        track-id="export_window"
        type="button"
      >
        <app-icon
          name="file-down"
        />
      </button>
      
      <button
        class="url-box__button"
        nz-tooltip=""
        nztooltipplacement="bottom"
        track-id="reload_docs"
        type="button"
      >
        <app-icon
          name="refresh-ccw"
        />
      </button>
      
      <button
        class="url-box__button"
        track-id="show_docs"
        type="button"
      >
          
      </button>
    </div>
    <div
      class="send-request__wrapper"
    >
      <button
        class="app-button active-primary send-request__button"
        track-id="set_request"
      >
        
        <span>
          
        </span>
        
      </button>
      
      
      <nz-dropdown-menu>
        <ul
          nz-menu=""
        >
          
          <li
            nz-menu-item=""
          >
              
          </li>
          
        </ul>
        
      </nz-dropdown-menu>
    </div>
  </div>
  
</app-url-box>
`;

exports[`UrlBoxComponent should render correctly with queryOperations > 1 1`] = `
<app-url-box>
  <div
    class="url-box"
  >
    <div
      class="url-box__method-wrapper"
    >
      <button
        class="url-box__button url-box__button--method"
        nz-dropdown=""
        nztrigger="click"
        track-id="http_verb"
        type="button"
        value="POST"
      >
         POST 
      </button>
      
      <nz-dropdown-menu>
        <ul
          nz-menu=""
        >
          <li
            nz-menu-item=""
            value="0"
          >
             POST 
          </li>
          
          <li
            nz-menu-item=""
            value="0"
          >
             GET 
          </li>
          
          <li
            nz-menu-item=""
            value="0"
          >
             PUT 
          </li>
          
          <li
            nz-menu-item=""
            value="0"
          >
             DELETE 
          </li>
          
          
        </ul>
        
      </nz-dropdown-menu>
    </div>
    <div
      class="url-box__input-container"
    >
      
      <app-x-input
        class="url-box__input"
        placeholder="Enter URL"
        track-id="set_url"
      />
      
    </div>
    <div
      class="url-box__input-actions"
    >
      
      <button
        class="url-box__button"
        nz-tooltip=""
        nztooltipplacement="bottom"
        track-id="save_to_collection"
        type="button"
      >
        <app-icon
          name="plus-circle"
        />
      </button>
      
      <button
        class="url-box__button"
        nz-tooltip=""
        nztooltipplacement="bottom"
        track-id="export_window"
        type="button"
      >
        <app-icon
          name="file-down"
        />
      </button>
      
      <button
        class="url-box__button"
        nz-tooltip=""
        nztooltipplacement="bottom"
        track-id="reload_docs"
        type="button"
      >
        <app-icon
          name="refresh-ccw"
        />
      </button>
      
      <button
        class="url-box__button"
        track-id="show_docs"
        type="button"
      >
          
      </button>
    </div>
    <div
      class="send-request__wrapper"
    >
      
      <button
        class="app-button active-primary send-request__button"
        nz-dropdown=""
        nzplacement="bottomRight"
        nztrigger="click"
        track-id="send_request"
      >
        <span>
          
        </span>
        
        
        
        
        
        
        <app-icon
          name="chevron-down"
        />
      </button>
      
      
      <nz-dropdown-menu>
        <ul
          nz-menu=""
        >
          <li
            nz-menu-item=""
          >
             operation 1 
          </li>
          
          <li
            nz-menu-item=""
          >
             operation 2 
          </li>
          
          
          <li
            nz-menu-item=""
          >
              
          </li>
          
        </ul>
        
      </nz-dropdown-menu>
    </div>
  </div>
  
</app-url-box>
`;
